atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/transaction/impl/SnapshotTransactionManager.java

```diff
@@ -170,13 +170,10 @@ protected boolean shouldStopRetrying(int numTimesFailed) {
     public <T, C extends PreCommitCondition, E extends Exception> T runTaskWithConditionThrowOnConflict(
             C condition, ConditionAwareTransactionTask<T, C, E> task) throws E, TransactionFailedRetriableException {
         checkOpen();
-        try {
-            OpenTransaction openTransaction = runTimed(
-                    () -> Iterables.getOnlyElement(startTransactions(ImmutableList.of(condition))), "setupTask");
-            return openTransaction.finish(transaction -> task.execute(transaction, condition));
-        } finally {
-            condition.cleanup();
-        }
+        OpenTransaction openTransaction =
+                runTimed(() -> Iterables.getOnlyElement(startTransactions(ImmutableList.of(condition))), "setupTask");
+        return openTransaction.finishWithCallback(
```

Comment:
This is different - previously, if the openTransaction L174 threw, cleanup was called. In the new, this is not the case. We can solve this by something like 

```
 checkOpen();
        OpenTransaction openTransaction;
        try {
            openTransaction = runTimed(
                    () -> Iterables.getOnlyElement(startTransactions(ImmutableList.of(condition))), "setupTask");
        } catch (Exception e) {
            condition.cleanup();
            throw e;
        }
        return openTransaction.finishWithCallback(
                transaction -> task.execute(transaction, condition), condition::cleanup);
```

at the expense of being less readable. We do not explicitly guarantee that cleanup will run

Reply:
Yeah, we don't. It is best effort, but I think we should (and what you describe looks right): startTransactions does involve an RPC to TimeLock, so there's a pretty real probability that that fails.